### ISSUE TITLE
Relation, IndirectRelation, and IndirectlyComparable are largely meaningless

### ISSUE THREAD

**Issue by CaseyCarter** [ericniebler/stl2#610]
[P1248](https://wg21.link/p1248) was a bit of a hatchet job. It ripped out `Relation`'s `CommonReference` requirement and the syntactic requirements on the common reference type - which is fine - but also all of the semantic requirements that held the concept together so it had meaning. Now `Relation` is:
```c++
template<class R, class T, class U>
  concept Relation =
    Predicate<R, T, U> && Predicate<R, U, T> &&
    Predicate<R, T, T> && Predicate<R, U, U>;
```
without any semantic relationships between the expressions `r(t, u)`, `r(u, t)`, `r(t1, t2)`, and `r(u1, u2)`. It simply requires those four seemingly-unrelated expressions to be valid. We fixed `StrictWeakOrder` by breaking `Relation`, `IndirectRelation`, and `IndirectlyComparable`. (To be clear, I'm not laying this on @tomaszkam - I had concerns and didn't make time to investigate. This is on me.)

We need to audit the uses of `IndirectRelation` and `IndirectlyComparable` - `Relation` is used only by those two concepts and `StrictWeakOrder` - and reestablish some semantics for those concepts.

We have to fix `Relation`, `IndirectRelation`, and `IndirectlyComparable` for C++20; shipping broken concepts is not an option.

**Comment by tomaszkam** [comment]
I will disagree with the view on the `Relation` and `IndirectRelation` - the relation in the mathematics means any pairing of the values of two domains (note that function is a relation). This is seven evident in case of the concepts model, as `Relation` is used as the base for `StrictWeakOrder` (that is _transtive_, _asymetric_ and _irreflexible_) and `IndirectlyComparable` (that is _transtive_, _symetric_ and _reflexible_), so the only common requirement is transitive. We can rename this concept to `TranstiveRelation` and require transitivity - currently, it only requires a set of expression that is required to check properties of orderings and equality (determine if given functor is specific kind of relation over specific set of arguments).

In my opinion, the `IndirectlyComparable` concept is as meaningful as before. The removed elements of `Relation` concept were not adding any semantics requirements. The only guarantee I have removed is that there is some C (maybe `variant<T, U>`) that `r(x, y)` is equivalent to `r(C(x), C(y))`, so we get four unrelated pairs of valid expressions, instead of four expressions.

You may want to give meaning for `IndirectlyComparable` by requiring it to model weak equivalence, but this would require the operation to be symmetric, transitive and reflexive, not `CommonReference`. My paper has not removed such properties for `Relation`, it just made this problem more glaring. In other words, the `Relation`, `IndirectRelation`, and `IndirectlyComparable` are and were meaningless, but don't blame my paper on it.

**Comment by tomaszkam** [comment]
To be honest, I am not really following how adding some type `C` that can be constructed from `T` and `U`, and `r(x, y) = r(C(x), C(y))` have added any semantics to the `IndirectlyComparable`.

**Comment by tomaszkam** [comment]
Finally, given the other linked issue, I think that the `IndirectlyComparable` and `IndirectRelation` concepts are currently overconstraining the algorithm - to realize the described effects of algorithms we need only `IndirectBianaryPredicate`, and obviously we have broken intended usages (calendar example).

Given that, I think that `IndirectlyComparable` may ended up meaningless on purpose, as from algorithms perspective, it is overconstrained version (requires all four combinations) of `IndirectBianaryPredicate`.

**Comment by tomaszkam** [comment]
> P1248 was a bit of a hatchet job. It ripped out Relation's CommonReference requirement and the syntactic requirements on the common reference type - which is fine - but also all of the semantic requirements that held the concept together so it had meaning.

Regarding this topic, could you explain how the following requirements:
```
    template<class R, class T, class U>
    concept Relation =
      Predicate<R, T, T> && Predicate<R, U, U> &&
      CommonReference<const remove_reference_t<T>&, const remove_reference_t<U>&> &&
      Predicate<R,
        common_reference_t<const remove_reference_t<T>&, const remove_reference_t<U>&>,
        common_reference_t<const remove_reference_t<T>&, const remove_reference_t<U>&>> &&
      Predicate<R, T, U> && Predicate<R, U, T>;

      Let
            r be an expression such that decltype((r)) is R,
            t be an expression such that decltype((t)) is T,
            u be an expression such that decltype((u)) is U, and
            C be common_reference_t<const remove_reference_t<T>&, const remove_reference_t<U>&>
        Relation<R, T, U> is satisfied only if
            bool(r(t, u)) == bool(r(C(t), C(u))).
            bool(r(u, t)) == bool(r(C(u), C(t))).

```

From my reading, there were guarantee that:
```
bool(r(t, t)) == bool(r(C(t), C(t)))
bool(r(u, u)) == bool(r(C(u), C(u)))
bool(r(t, u)) == bool(r(C(t), C(u)))
bool(r(u, t)) == bool(r(C(u), C(t)))
```
, but it does not define any relation between values of `r(t,u)`, `r(u, t)` or `r(t, t)`, i.e. there was not tied together. `StrictWeakOrdering` is giving coupling by requiring it to be transtivie, assymetric and irreflexible.

**Comment by CaseyCarter** [comment]
To be clear: I'm not suggesting that we must return the prior semantics - only that these concepts should have *some* semantics. "These four unrelated `Boolean` expressions must be valid and equality preserving" is not semantically useful without placing any requirements on their relative values or domains. At best it's simply a bundle of syntax that saves some typing in "real" concepts that refine `Relation` which *do* enforce such semantics.

> the relation in the mathematics means any pairing of the values of two domains (note that function is a relation)

A relation is a subset of the ordered tuples in the cartesian product of some number of sets. `Predicate<F, T, U>` models the mathematical notion of a relation. Neither the prior nor current formulations of `Relation` do so - at the very least the naming is questionable.

> The removed elements of `Relation` concept were not adding any semantics requirements.

The fact that `CommonRefence` requires the mappings from `T` and `U` to `C` (`common_reference_t<const remove_reference_t<T>&, const remove_reference_t<U>&>`) to be 1-to-1 and that the constituent expressions were required to be equivalent to the expressions on `C` established relationships between the constituent expressions of `Relation`. For example, if `C(t)` equals `C(u)` then `bool(r(t, t)) == bool(r(C(t), C(t)))` [axiomatic], `r(C(t), C(t))` equals `r(C(u), C(u))` [`Predicate` requires equality-preservation], and `bool(r(C(u), C(u))) == bool(r(u, u))` [axiomatic]. (Although since we failed to require the domains to be the same there was no guarantee that all of these expressions are well-defined when any of them are.) `C(t)` equal to `C(u)` similarly implied that `bool(r(t, t))`,  `bool(r(t, u))`,  `bool(r(u, t))`, and `bool(r(u, u))` are all equal, which relates all four of the required expressions.

TLDR: the prior requirements established that `r` is type-independent: `bool(r(x1, y1)) == bool(r(x2, y2))` must hold when `x1` and `x2` represent the same platonic value and `y1` and `y2` represent the same platonic value regardless of the types of `x1`, `x2`, `y1`, and `y2`.

> You may want to give meaning for `IndirectlyComparable` by requiring it to model weak equivalence...
> My paper has not removed such properties for `Relation`, it just made this problem more glaring.

Very true: `IndirectRelation` and `IndirectlyComparable` were broken before, and I hadn't noticed until P1248 made that fact more apparent.

**Comment by tomaszkam** [comment]
> TLDR: the prior requirements established that r is type-independent: bool(r(x1, y1)) == bool(r(x2, y2)) must hold when x1 and x2 represent the same platonic value and y1 and y2 represent the same platonic value regardless of the types of x1, x2, y1, and y2.

I think that this conclusion is not valid, because there is nothing (I believe) that would prevent `C` from being a sum type (`variant<T, U>`) regardless of types relation (i.e. even for related types like `std::string` and `std::string_view`) - in that situation the fact that `t` and `u` represent the same platonic value will not imply that `r(t, u)` is equal to `r(u,t)` because `C(t) != C(u)`. [ Only requirement of `CommonReference` is that is it equality preserving in a domain of single type, i.e. `t1 == t2` implies `C(t1) == C(t2)` ]

I am bringing this up, because the existence of some common platonic value for `T` and `U` could be assumed only for equality, not equivalence (same argumentation as in my paper). And for equivalence (`Employee` and `std::string_view`) the `C` would usually be and sum type, so not providing any real additional semantic.

**Comment by CaseyCarter** [comment]
> there is nothing that would prevent C from being a sum type (variant<T, U>) regardless of types relation (i.e. even for related types like std::string and std::string_view) 

You tell the library that `x1` and `x2` represent the same platonic value by providing a common type / common reference type such that `C(x1)` is equal to `C(x2)`. If you use `variant<string, string_view>` as the common type of `string` and `string_view`, you're asserting that `string` and `string_view` values never represent the same platonic value. That isn't completely insane, but is inconsistent with the cross-type operators the library defines for those types so you couldn't satisfy `EqualityComparableWith` or `StrictTotallyOrderedWith`. (We forbid specializing `common_type` and `basic_common_reference` for two types neither of which depend on a user-defined type so users can't create weirdness like this.)

> the existence of some common platonic value for `T` and `U` could be assumed only for equality, not equivalence

I think you mean "common type" here, not "common value". And yes, I agree. I'm not suggesting that we need to return the common type / reference semantics, I'm pointing out how they linked the required expressions of `Relation` together into something coherent. We don't want to bring back the old meaning of the concept, but it needs to have *some* useful meaning or we should yank it from the working draft.

**Comment by tomaszkam** [comment]
>  And yes, I agree. I'm not suggesting that we need to return the common type / reference semantics, I'm pointing out how they linked the required expressions of Relation together into something coherent.

I am approaching this from the other point here, we both have agreed that in mathematics the _relation_ is nothing more than a simple `Predicate<F, X, Y>` and there is nothing more required (i.e. it is not more coherent than that). There is nothing that is linking the expression, and I think we should not invent something for concepts on the fly.

The problem with the representation as `Predicate<F, X, Y>` is that the syntactic requirement guaranteed by this concept does not allow you to validate properties of the relation. For example, checking the _symmetry_, requires also `F` to meet `Predicate<F, Y, X>`. And checking the _reflexibility_ requires both `Predicate<F, X, X>` and `Predicate<F, Y, Y>`. This leads us to the set of operations that is required for `F` to be checked/validated against the axioms of specialization relations (like ordering/equivalence) and I am fine with naming this _checkable relation_ concept just `Relation` (because when we say about relations in WD, we usually mean orderings, not functions).
Note that this still does not have any requirements on relations between the values of `r(x, x)`, `r(x, y)`, `r(y, x)`, and `r(y,y)`.

Now we refine this concept into `StrictWeakOrder`, that provides a constrains on values (give coherency), by saying that `r` must be weak ordering. We could also refine it into `Equivalence`/`IndirectEquivalence` meaning equivalence.

The whole point is that I am not seeing the `Relation` itself as being meaningless, contrary I am seeing it to being a right concept. However, I agree that usage of `InderectRelation` needs to be changed to either `IndirectEquivalence` (or `IndirectlyComparable` if we say it is equivalence) or `IndirectBinaryPrediciate`. I think that we should do the later (see the other discussions).

**Comment by tomaszkam** [comment]
> I'm not suggesting that we need to return the common type / reference semantics, I'm pointing out how they linked the required expressions of Relation together into something coherent.

And I was responding, that the `CommonReference` requirement is too weak to provide the guarantee you were describing. As long as you allow `C` to be `variant<T, U>`, the `CommonReference` additional requirements are meaningless. I am well aware that it is not an intent of this requirement and the standard library definition are not behaving that way, but it seems that the `CommonReference` concept miss some semantic requirement.

**Comment by tomaszkam** [comment]
[P1716: `ranges` compare algorithm are over-constrained ](https://wg21.link/p1716) aims to address the problem.

**Comment by ericniebler** [comment]
@tomaszkam It looks like P1716 was forwarded by LEWG for C++20, but LWG didn't have time to see it in Cologne, is that correct? We may need an NB comment for this.

**Comment by JohelEGP** [comment]
Looking at https://github.com/cplusplus/papers/issues/482, so it would seem.

**Comment by tomaszkam** [comment]
@ericniebler Yes, the P1716 did make it through LWG in Colonge. I plan to submit an LWG issue for the problem addressed and will link it here.

**Comment by ericniebler** [comment]
> Yes, the P1716 did make it through LWG in Colonge.

I think you either mean "did _not_", or else you mean L<i>E</i>WG.

**Comment by tomaszkam** [comment]
Actually both, was accepted by LEWG, LWG has not managed to review it.

**Comment by CaseyCarter** [comment]
> I plan to submit an LWG issue for the problem addressed and will link it here.

@tomaszkam Please do so we can close this issue as superseded by the LWG issue.

**Comment by ericniebler** [comment]
I seem to recall am email to lwgchair about this, and the response was that an issue was inappropriate. I think this needs an NB comment.
